sys/random/init.c

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2019 Freie Universität Berlin
+ *               2020 Sören Tempel <tempel@uni-bremen.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <stdint.h>
+
+#include "log.h"
+#include "random.h"
+
+#ifdef MODULE_PUF_SRAM
+#include "puf_sram.h"
+#endif
+#ifdef MODULE_PERIPH_HWRNG
+#include "periph/hwrng.h"
+#endif
+#ifdef MODULE_PERIPH_CPUID
+#include "luid.h"
+#endif
+
+#define ENABLE_DEBUG (1)
+#include "debug.h"
+
+void auto_init_random(void)
+{
+    uint32_t seed = 1;
+#ifdef MODULE_PUF_SRAM
+    /* TODO: hand state to application? */
+    if (puf_sram_state) {
+        LOG_WARNING("random: PUF SEED not fresh\n");
+    }
+    seed = puf_sram_seed;
+#elif defined (MODULE_PERIPH_HWRNG)
+    hwrng_read(&seed, 4);
+#elif defined (MODULE_PERIPH_CPUID)
+    luid_get(&seed, 4);
+#else
+    LOG_WARNING("random: USING COMPILE-TIME GENERATED SEED!\n");
+#endif
+    seed *= RANDOM_SEED_DEFAULT;
```

Comment:
Not sure if multiplication is really the best mathematical operation to use here. Maybe it makes sense to use XOR instead? Any preferences on this?

Reply:
Not a crypto crack either, but intuitively, XOR (which "shakes up the bits") seems more suitable than multiplication (which if it is "accidentally" of two can lead to a lot of LSB zeros).

Reply:
This is only a compile-static fallback, right? I would rather leave it in the `#else` case and with that I don't see the need for combing the static seed with `seed`  anyway.  In the long announced and yet not seen entropy module (shame on me) that I plan to propose, a static input will be enabled as entropy source which will then (ideally) be hashed together with other entropy sources.

Reply:
Is not strictly necessary to combine the seed with `RANDOM_SEED_DEFAULT` unconditionally. Preferably I would not want to recompile the file using `RANDOM_SEED_DEFAULT` on each make invocation if some other method of generating random numbers is available though I didn't find any easy way to detect whether or not this is the case from the `Makefile`. If there is such a method let me know, in that case it would be possible to only mark the file as `.PHONY` if no run-time random number generation method is available.